mm2src/coins/utxo/utxo_common.rs

```diff
@@ -550,82 +540,146 @@ impl<'a, T: AsRef<UtxoCoinFields> + UtxoTxGenerationOps> UtxoTxBuilder<'a, T> {
+            .outputs
+            .iter()
+            .fold(0u64, |required, output| required + output.value);
         match self.fee_policy {
             FeePolicy::SendExact => {
-                let mut outputs_plus_fee = self.sum_outputs_value + self.tx_fee;
-                if self.sum_inputs >= outputs_plus_fee {
-                    self.change = self.sum_inputs - outputs_plus_fee;
-                    if self.change > self.dust() {
-                        // there will be change output
-                        if let ActualTxFee::Dynamic(ref f) = actual_tx_fee {
-                            self.tx_fee += (f * P2PKH_OUTPUT_LEN) / KILO_BYTE;
-                            outputs_plus_fee += (f * P2PKH_OUTPUT_LEN) / KILO_BYTE;
-                        }
-                    }
-                    if let Some(min_relay) = self.min_relay_fee {
-                        if self.tx_fee < min_relay {
-                            let fee_diff = min_relay - self.tx_fee;
-                            outputs_plus_fee += fee_diff;
-                            self.tx_fee += fee_diff;
-                        }
-                    }
-                    self.sum_inputs >= outputs_plus_fee
-                } else {
-                    false
-                }
+                sum_output += self.total_tx_fee();
             },
-            FeePolicy::DeductFromOutput(_) => {
-                if self.sum_inputs >= self.sum_outputs_value {
-                    self.change = self.sum_inputs - self.sum_outputs_value;
-                    if self.change > self.dust() {
-                        if let ActualTxFee::Dynamic(ref f) = actual_tx_fee {
-                            self.tx_fee += (f * P2PKH_OUTPUT_LEN) / KILO_BYTE;
-                        }
-                    }
-                    if let Some(min_relay) = self.min_relay_fee {
-                        if self.tx_fee < min_relay {
-                            self.tx_fee = min_relay;
-                        }
+            FeePolicy::DeductFromOutput(_) => {},
+        };
+        sum_output
+    }
+
+    fn add_tx_inputs(&mut self, amount: u64) -> u64 {
+        self.tx.inputs.clear();
+        let mut total = 0u64;
+        for utxo in self.required_inputs.clone() {
+            self.tx.inputs.push(UnsignedTransactionInput {
+                previous_output: utxo.outpoint,
+                prev_script: utxo.script,
+                sequence: SEQUENCE_FINAL,
+                amount: utxo.value,
+            });
+            total += utxo.value;
+        }
+
+        for utxo in self.available_inputs.clone() {
+            if total >= amount {
+                break;
+            }
+            self.tx.inputs.push(UnsignedTransactionInput {
+                previous_output: utxo.outpoint,
+                prev_script: utxo.script,
+                sequence: SEQUENCE_FINAL,
+                amount: utxo.value,
+            });
+            total += utxo.value;
+        }
```

Comment:
This should be slightly cheaper:

```suggestion
        for utxo in &self.required_inputs {
            self.tx.inputs.push(UnsignedTransactionInput {
                previous_output: utxo.outpoint,
                prev_script: utxo.script.clone(),
                sequence: SEQUENCE_FINAL,
                amount: utxo.value,
            });
            total += utxo.value;
        }

        for utxo in &self.available_inputs {
            if total >= amount {
                break;
            }
            self.tx.inputs.push(UnsignedTransactionInput {
                previous_output: utxo.outpoint,
                prev_script: utxo.script.clone(),
                sequence: SEQUENCE_FINAL,
                amount: utxo.value,
            });
            total += utxo.value;
        }
```

Reply:
done in 50a38f800ee3eebc910d27b3ebd63e599ef33c6c